fastqpreprocessing/src/fastqprocess.cpp

```diff
@@ -0,0 +1,428 @@
+     for (int i = 0; i < options.R1s.size(); i++) {
+        readers[i].join();
+     }
+     // set the stop flag for the writers
+     samrecord_data->stop = true;
+
+     // ask the writers to make one more loop in the whilte loop
+     for (int j = 0; j < samrecord_data->num_files; j++) {
+        if (sem_post(&semaphores[j]) == -1)
+           error("sem_post: semaphores");
+     }
+
+     // wait for the writers to stop after they have seen the stop flag
+     for (int i = 0; i < samrecord_data->num_files; i++) {
+        writers[i].join();
+     }
+
+     // destroy the semaphores
+     for (int i = 0; i < samrecord_data->num_files; i++) {
+         sem_destroy(&semaphores[i]);
+     }
+
+     // destroy the semaphores for semaphores_workers
+     for (int i = 0; i < samrecord_data->num_files; i++) {
+         sem_destroy(&semaphores_workers[i]);
+     }
+     
+     // delete reader and writer threads
+     delete [] readers;
+     delete [] writers;
+}
+
+/** @copydoc bam_writers */
+void bam_writers(int windex, SAM_RECORD_BINS *samrecord_data) {
+    SamFile samOut;
+    std::string outputfile;
+    
+    // name of the output file
+    char buf[MAX_FILE_LENGTH];
+    sprintf(buf, "subfile_%d.bam", windex);
+    outputfile = buf;
+
+    // open to write the outputfile
+    samOut.OpenForWrite(outputfile.c_str());
+
+    // Write the sam header.
+    SamFileHeader samHeader;
+
+    // add the HD tags for the header
+    samHeader.setHDTag("VN", "1.6");
+    samHeader.setHDTag("SO", "unsorted");
+
+    // add the RG group tags
+    SamHeaderRG *headerRG = new SamHeaderRG;
+    headerRG->setTag("ID", "A");
+    headerRG->setTag("SM", samrecord_data->sample_id.c_str());
+    samHeader.addRG(headerRG);
+   
+    // add the header to the output bam
+    samOut.WriteHeader(samHeader);
+
+    // keep writing foever, until there is a flag to stop
+    while (true) {
+      // wait untile some data is ready from a reader thread
+      if (sem_wait(&semaphores[windex]) == -1)
+         error("sem_wait:semaphores");
+
+      // write out the record buffers for the reader thread "active_thread_no"
+      // that signalled that buffer is ready to be written
+      SamRecord *samRecord  = \
+         samrecord_data->samrecords[samrecord_data->active_thread_no];
+      // go through the index of the samrecords that are stored for the current 
+      // writer, i.e., "windex" or the corresponding BAM file
+      for (auto index : samrecord_data->file_index[samrecord_data->active_thread_no][windex]) {
+           samOut.WriteRecord(samHeader, samRecord[index]);
+      }
+
+      // lets the reads thread know that I am done writing the 
+      // buffer that are destined to be my file
+      if (sem_post(&semaphores_workers[windex]) == -1)
+          error("sem_post: semaphores_workers");
+
+      // time to stop variable is valid
+      if (samrecord_data->stop) break;
+    }
+
+    // close the bamfile
+    samOut.Close();
```

Comment:
I would not! Because,  each thread only writes to it own file, i.e., samOut and the only way it get to it from the infinite while loop is by exiting via the if (samrecord_data->stop) break (means nothing more to write) ;, which is just a couple of lines  above